### DISCUSSION TITLE
Convert to PDF dose not worked in my docker (marp-cli)

### DISCUSSION THREAD

**Discussion by Uno-Takashi** [marp-team/marp#165] [Q&A] [answered]
I prepared the following simple docker environment and tried to run .md to .pdf in it.

```
FROM node:16

WORKDIR /workspaces
RUN apt update

# install chrome
RUN curl -LO https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
RUN apt-get install -y ./google-chrome-stable_current_amd64.deb
RUN rm google-chrome-stable_current_amd64.deb 

RUN npm install -g @marp-team/marp-cli
```

I was able to convert markdown to html correctly, but the conversion to pdf has the following error.

```
npx @marp-team/marp-cli@latest contents/main.md --pdf

Need to install the following packages:
  @marp-team/marp-cli@latest
Ok to proceed? (y) y
[  INFO ] Converting 1 markdown...
[ ERROR ] Failed converting Markdown. (Protocol error (Target.setDiscoverTargets): Target closed.)
```

I have confirmed that the same phenomenon occurs when I change the base image to ubuntu. Similarly, changing google-chrome-stable to edge and chromium causes the same problem.

If you don't install chrome, you will get "WORN", so it is assumed that OS and marp-cli recognizes chrome.

In this [issue](https://github.com/marp-team/marp-cli/issues/219), it is mentioned that the problem is simply due to the fact that chrome is not installed, but we believe that this is a more general problem.

Why is marp-cli not working in this environment?

By the way, [here's how it failed](https://github.com/Uno-Takashi/Marp-on-VSCode/actions/runs/1221393577) when run with github actions.

**Comment by yhatt** [comment, marked as answer]
Marp CLI is internally using Puppeteer, and it is known as difficult to get working within Docker container. Follow Puppeteer's guide for Docker:
https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#running-puppeteer-in-docker

Or find out similar issues from Puppeteer repository: https://github.com/puppeteer/puppeteer/search?q=setDiscoverTargets+docker&type=issues

**Reply by yhatt** [reply]
**UPDATE**: The latest Marp CLI v1.4.1 has improved executing Marp CLI in the container image by marp-team/marp-cli#389. Now CLI detects whether it is running on Docker image, and try to execute Chromium browser with better settings that matched to the current environment.

**Reply by yhatt** [reply]
**UPDATE:** It may have resolved in [Marp CLI v1.4.2](https://github.com/marp-team/marp-cli/releases/tag/v1.4.2).

- https://github.com/marp-team/marp-cli/issues/395
- https://github.com/marp-team/marp-cli/pull/400
